### PR TITLE
Fix allocation and health checks in IPv6 only mode

### PR DESCRIPTION
**Pull request by tgraf**
Several hard dependencies on IPv4 have found their way into the codebase. Allow running in IPv6 mode again without allocating any IPv4 addresses.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6710)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

File: daemon/daemon.go [hunk 1]
```diff
@@ -1049,6 +1042,13 @@ func NewDaemon(dp datapath.Datapath) (*Daemon, *endpointRestoreState, error) {
 	log.Info("Initializing IPAM")
 	ipam.Init()
 
+	wOpts := map[string]string{
```
Comment (aanm) [inline daemon/daemon.go:1053]: Add a comment stating this can not be moved before the IPAM initialization

File: pkg/ipam/allocator.go [hunk 2]
```diff
@@ -79,34 +80,54 @@ func (ipam *IPAM) AllocateIPString(ipAddr string) error {
...
-		ipConf, err := ipam.IPv6Allocator.AllocateNext()
+		ipv6, err = ipam.AllocateNextFamily(IPv6)
 		if err != nil {
-			return nil, nil, err
+			return
 		}
 
-		ipv6 = ipConf
-		metrics.IpamEvent.WithLabelValues(metricAllocate, familyIPv6).Inc()
 	}
```
Comment (aanm) [inline pkg/ipam/allocator.go:122]: in this line it needs a :
```
defer func(){
 if err != nil {
    ipam.ReleaseIPv6Address(ipv6)
  }
}
```

File: pkg/ipam/ipam.go [hunk 3]
```diff
@@ -49,52 +47,15 @@ const (
...
-	ipam.IPAMConfig.Routes = append(ipam.IPAMConfig.Routes,
-		// IPv4
-		&cniTypes.Route{
-			Dst: net.IPNet{
-				IP:   nodeAddressing.IPv4().Router(),
-				Mask: net.CIDRMask(32, 32),
-			},
-		},
-		&cniTypes.Route{
-			Dst: defaults.IPv4DefaultRoute,
```
Comment (aanm) [inline pkg/ipam/ipam.go]: Where is the gateway defined now? Does this means that `github.com/containernetworking/plugins/plugins/ipam/host-local/backend/allocator` is not used anywhere?

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/ipam/ipam.go (hunk 3 above)
Comment (tgraf) [reply pkg/ipam/ipam.go]: I didn't find any code using this. I will double check again.

File: pkg/ipam/allocator.go (hunk 2 above)
Comment (tgraf) [reply pkg/ipam/allocator.go:122]: This is an existing bug. I'll fix it in a separate so we can backport.

Comment (tgraf) [reply pkg/ipam/allocator.go:122]: https://github.com/cilium/cilium/pull/6719

File: pkg/ipam/ipam.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+type ErrAllocation error
+
+// Family is the type describing all address families support by the IP
+// allocation manager
+type Family string
+
+const (
+	IPv6 Family = "ipv6"
+	IPv4 Family = "ipv4"
```
Comment (joestringer) [inline pkg/ipam/ipam.go:45]: These appear to duplicate `familyIPv6` and `familyIPv4`.

File: daemon/daemon.go [hunk 5]
```diff
@@ -1123,16 +1123,25 @@ func NewDaemon(dp datapath.Datapath) (*Daemon, *endpointRestoreState, error) {
...
+			d.nodeDiscovery.localNode.IPv4HealthIP = health4
+			log.Debugf("IPv4 health endpoint address: %s", health4)
 		}
 
-		d.nodeDiscovery.localNode.IPv4HealthIP = health4
-		log.Debugf("IPv4 health endpoint address: %s", health4)
+		if option.Config.EnableIPv6 {
+			health6, err := d.ipam.AllocateNextFamily(ipam.IPv6)
+			if err != nil {
+				return nil, restoredEndpoints, fmt.Errorf("unable to allocate health IPs: %s,see https://cilium.link/ipam-range-full", err)
```
Comment (joestringer) [inline daemon/daemon.go:1145]: There's no de-allocation of the potentially allocated `health4` address in this error path. (Though admittedly it's probably less likely you'll run out of IPv6 addresses than IPv4 addresses and actually hit this..)

**Review by joestringer — CHANGES_REQUESTED** [review]
I posted few minor comments below, one unlikely bug and one duplicate set of symbols.

I also noticed that the node element model expects IPv4 AND IPv6 addresses, which is a softer dependency not handled in this PR:
https://github.com/cilium/cilium/blob/17f945350af127ed58aa105487140509fad7d646/pkg/node/node.go#L199-L221

File: pkg/ipam/ipam.go (hunk 4 above)
Comment (tgraf) [reply pkg/ipam/ipam.go:45]: Yes, I did not want to export untyped constant and I didn't want to refactor all code either and keep the changes small.

File: daemon/daemon.go (hunk 5 above)
Comment (tgraf) [reply daemon/daemon.go:1145]: Yes, this is true for all allocations and a lot of other resources. It is currently assumed that if `NewDaemon()` fails, then the agent process is quit. I can fix it up but it doesn't really solve the overall problem.

**Comment by tgraf** [comment]
> I also noticed that the node element model expects IPv4 AND IPv6 addresses, which is a softer dependency not handled in this PR:

@joestringer That code seems to always provide IPv4 only if IPv4 is enabled and otherwise provides IPv6. I'm not sure why we don't just provide both depending on availability. Do you see anything that needs to change in this PR?

**Comment by tgraf** [comment]
test-me-please

**Comment by joestringer** [comment]
> @joestringer That code seems to always provide IPv4 only if IPv4 is enabled and otherwise provides IPv6. I'm not sure why we don't just provide both depending on availability. Do you see anything that needs to change in this PR?

My read was that if either is disabled, it will provide nothing (specifically the lines below), which would remove that information from the API status response, perhaps leading to the unintended complete disabling of cilium-health pinging to those nodes (rather than just disabling the protocol that's disabled)?

https://github.com/cilium/cilium/blob/17f945350af127ed58aa105487140509fad7d646/pkg/node/node.go#L200-L202

That said, this is presumably also a bug again independent of this PR so shouldn't really block this one.

**Review by joestringer — APPROVED** [review]

**Comment by tgraf** [comment]
test-me-please

File: pkg/ipam/ipam.go (hunk 3 above)
Comment (tgraf) [reply pkg/ipam/ipam.go]: Removed the vendor code

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

Glitch in vagrant:
```
00:41:50.765 ==> k8s2-1.13: Checking for guest additions in VM...
00:41:50.765 ==> k8s2-1.13: Setting hostname...
00:41:50.765 ==> k8s2-1.13: Configuring and enabling network interfaces...
00:46:39.703 Sending interrupt signal to process
00:46:43.691 script returned exit code 255
```

**Comment by tgraf** [comment]
test-me-please

```
hudson.plugins.git.GitException: Failed to fetch from http://github.com/cilium/cilium
	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:888)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1155)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1186)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:120)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:90)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:77)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution$1$1.call(SynchronousNonBlockingStepExecution.java:51)
	at hudson.security.ACL.impersonate(ACL.java:290)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution$1.run(SynchronousNonBlockingStepExecution.java:48)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by tgraf** [comment]
test-me-please

**Pull request merged**